CMakeLists.txt

```diff
@@ -1825,23 +1825,48 @@ elseif(APPLE)
   endif()
 
   if(SDL_JOYSTICK)
+    file(GLOB MFI_JOYSTICK_SOURCES ${SDL2_SOURCE_DIR}/src/joystick/iphoneos/*.m)
     if(IOS OR TVOS)
-      file(GLOB JOYSTICK_SOURCES ${SDL2_SOURCE_DIR}/src/joystick/iphoneos/*.m ${SDL2_SOURCE_DIR}/src/joystick/steam/*.c)
-      set(SDL_JOYSTICK_MFI 1)
+      file(GLOB JOYSTICK_SOURCES ${SDL2_SOURCE_DIR}/src/joystick/steam/*.c)
+      set(HAVE_FRAMEWORK_GAMECONTROLLER 1)
+      set(HAVE_FRAMEWORK_COREHAPTICS 1)
       if(IOS)
         set(SDL_FRAMEWORK_COREMOTION 1)
       endif()
-      set(SDL_FRAMEWORK_GAMECONTROLLER 1)
-      set(SDL_FRAMEWORK_COREHAPTICS 1)
       set(HAVE_SDL_SENSORS 1)
     else()
-      file(GLOB JOYSTICK_SOURCES ${SDL2_SOURCE_DIR}/src/joystick/darwin/*.c ${SDL2_SOURCE_DIR}/src/joystick/iphoneos/*.m)
-      # FIXME: add checks for SDL_JOYSTICK_MFI???
+      file(GLOB JOYSTICK_SOURCES ${SDL2_SOURCE_DIR}/src/joystick/darwin/*.c)
+      set_source_files_properties(${MFI_JOYSTICK_SOURCES} PROPERTIES COMPILE_FLAGS -fobjc-weak)
+      check_objc_source_compiles("
+          #include <AvailabilityMacros.h>
+          #include <TargetConditionals.h>
+          #import <Foundation/Foundation.h>
+          #import <GameController/GameController.h>
+          #if MAC_OS_X_VERSION_MIN_REQUIRED < 1080
+          #error GameController framework doesn't work on this configuration
+          #endif
+          #if TARGET_CPU_X86
+          #error GameController framework doesn't work on this configuration
+          #endif
+          int main() {}" HAVE_FRAMEWORK_GAMECONTROLLER)
+      check_objc_source_compiles("
+          #include <AvailabilityMacros.h>
+          #include <TargetConditionals.h>
+          #import <Foundation/Foundation.h>
+          #import <CoreHaptics/CoreHaptics.h>
+          int main() {}" HAVE_FRAMEWORK_COREHAPTICS)
       set(SDL_JOYSTICK_IOKIT 1)
       set(SDL_FRAMEWORK_IOKIT 1)
       set(SDL_FRAMEWORK_FF 1)
     endif()
-    set(SOURCE_FILES ${SOURCE_FILES} ${JOYSTICK_SOURCES})
+    if(HAVE_FRAMEWORK_GAMECONTROLLER)
```

Comment:
Do we not want both of `HAVE_FRAMEWORK_GAMECONTROLLER` and `HAVE_FRAMEWORK_COREHAPTICS` here?

Reply:
I don't think so. We can build the MFI backend without CoreHaptics support (see `ENABLE_MFI_RUMBLE` in `SDL_mfijoystick.m`).

Reply:
So the autotools check is more limiting?

Reply:
It looks like it. I assume `-Wl,-weak_framework,CoreHaptics` would fail if CoreHaptics.framework doesn't exist in the current SDK version, so we'd end up not enabling MFI at all.

The Autotools behavior may actually be beneficial though. With this change as is, if you built with an SDK that lacks CoreHaptics and then run on macOS 11, the MFI joystick implementation would take over but have no haptics or LED controls.

Reply:
Then why not match the cmake behavior to autotools?

Reply:
I will :)